utilities.py

```diff
@@ -171,10 +174,319 @@ def closest(start: list, test_pts: list):
 
     return closest_pt, distance_minimum
 
+
 def simulate_sensors(environment, sensors):
     '''Simulate a list of sensors'''
 
     ROBOT = environment.get("ROBOT", None)
     for d_id in sensors:
         if d_id in ROBOT.sensors:
             ROBOT.sensors[d_id].simulate(0, environment)
+
+
+def angle(segment1:list, segment2:list):    
+
+    theta1 = math.degrees(math.atan2(segment1[0][1]-segment1[1][1],segment1[0][0]-segment1[1][0]))
+    theta2 = math.degrees(math.atan2(segment2[0][1]-segment2[1][1],segment2[0][0]-segment2[1][0]))
+
+    diff = abs(theta1 - theta2)
+    
+    # while diff>=360:
+    #     diff-=360
+    
+    if diff>270:
+        return 360-diff
+    if diff>180:
+        return diff - 180
+    if diff>90:
+        return 180 - diff
+
+
+def closestFast(start: list, test_pts: list):
+    """
+    Returns the closest point in the test_pts list to the point start, and
+    the Euclidean distance between them.
+    """
+    
+    # If the list is empty, return the empty list and a nan for length
+    if not test_pts:
+        return test_pts, math.nan
+
+    # Otherwise calculate the closest point to the "start" point
+    else:
+        distSq_minimum = math.inf
+        closest_pt = []
+        for test_pt in test_pts:
+            # calculates squared distance between start and test_pt (to avoid sqrt calculation)
+            distSq = (test_pt[0] - start[0]) ** 2 + (test_pt[1] - start[1]) ** 2
+
+            # minimum squared distance corresponds to minimum distance
+            if distSq < distSq_minimum:
+                distSq_minimum = distSq
+                closest_pt = test_pt
+
+    #TODO: potentially only calculate sqrt once rather than 7 times per sensor
```

Comment:
Is this completed by the `closestFast()` function?

Reply:
It is now. Instead of storing the ray_lengths in the ultrasonic sensor object, I store ray_lengths_squared, and only squareroot the smallest one when returning the reading

Reply:
I am a little worried that performance improvements beyond this point would sacrifice readability, and even this one I'm not if the readability - performance gain trade-off is worth it. I didn't want to refactor too many things to make it more readable because my idea of readability might be different from someone else's